tls/s2n_cipher_suites.c

```diff
@@ -1155,24 +1155,138 @@ int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_C
     return 0;
 }
 
-static int s2n_wire_ciphers_contain(const uint8_t *match, const uint8_t *wire, uint32_t count, uint32_t cipher_suite_len)
+static int s2n_wire_ciphers_contain(const uint8_t *match, struct s2n_stuffer *wire_stuffer, uint32_t count, uint32_t cipher_suite_len)
 {
     for (uint32_t i = 0; i < count; i++) {
-        const uint8_t *theirs = wire + (i * cipher_suite_len) + (cipher_suite_len - S2N_TLS_CIPHER_SUITE_LEN);
+        uint8_t theirs[S2N_SSLv2_CIPHER_SUITE_LEN] = {0};
+        POSIX_GUARD(s2n_stuffer_read_bytes(wire_stuffer, theirs, cipher_suite_len));
 
         if (!memcmp(match, theirs, S2N_TLS_CIPHER_SUITE_LEN)) {
+            POSIX_GUARD(s2n_stuffer_reread(wire_stuffer));
             return 1;
         }
     }
 
+    POSIX_GUARD(s2n_stuffer_reread(wire_stuffer));
     return 0;
 }
 
+/* Finds and sets the index of the ChaCha20 cipher in the server security policy. If ChaCha20 does not exist, then the index is set to -1. */
+static S2N_RESULT s2n_get_index_of_chacha20_in_cipher_preferences(const struct s2n_cipher_preferences* cipher_preferences, ssize_t* idx) {
+    RESULT_ENSURE_REF(cipher_preferences);
+    RESULT_ENSURE_REF(idx);
+
+    const uint8_t chacha20_iana[S2N_TLS_CIPHER_SUITE_LEN] = { TLS_CHACHA20_POLY1305_SHA256 };
+
+    for (int i = 0; i < cipher_preferences->count; i++) {
+        const uint8_t *cipher_iana = cipher_preferences->suites[i]->iana_value;
+        if (s2n_constant_time_equals(chacha20_iana, cipher_iana, S2N_TLS_CIPHER_SUITE_LEN)) {
+            *idx = i;
+            return S2N_RESULT_OK;
+        }
+    }
+
+    *idx = -1;
+    return S2N_RESULT_OK;
+}
+
+/* If the client has ChaCha20 as their most preferred cipher suite (first cipher suite) then set flag to true. */
+static S2N_RESULT s2n_wire_indicates_chacha20_boosting(struct s2n_stuffer* wire_stuffer, uint8_t count, uint32_t cipher_suite_len, bool* chacha20_boosted) {
+    RESULT_ENSURE_REF(wire_stuffer);
+    RESULT_ENSURE_REF(chacha20_boosted);
+        
+    *chacha20_boosted = false;
+
+    if (count == 0) {
+        return S2N_RESULT_OK;
+    }
+
+    uint8_t clients_first_cipher_iana[S2N_SSLv2_CIPHER_SUITE_LEN] = {0};
+    RESULT_GUARD_POSIX(s2n_stuffer_read_bytes(wire_stuffer, clients_first_cipher_iana, cipher_suite_len));
+    RESULT_GUARD_POSIX(s2n_stuffer_reread(wire_stuffer));
+    const uint8_t chacha20_iana[S2N_TLS_CIPHER_SUITE_LEN] = { TLS_CHACHA20_POLY1305_SHA256 };
+
+    if (s2n_constant_time_equals(chacha20_iana, clients_first_cipher_iana, S2N_TLS_CIPHER_SUITE_LEN)) {
+        *chacha20_boosted = true;
+        return S2N_RESULT_OK;
+    }
+
+    return S2N_RESULT_OK;
+}
+
+/* Checks that a mutually-supported cipher suite can be used in this connection */
+static S2N_RESULT s2n_validate_cipher_suite_match(struct s2n_connection* conn, struct s2n_cipher_suite** potential_match, bool* match_is_valid) {
```

Comment:
Nit: even though we don't have a linter, try to match the style of the codebase.
```suggestion
/* Checks that a mutually-supported cipher suite can be used in this connection */
static S2N_RESULT s2n_validate_cipher_suite_match(struct s2n_connection *conn, struct s2n_cipher_suite **potential_match, bool *match_is_valid) {
```